File: algorithm/binary_tree/python/binary_tree_right_side.py
"""
Binary Tree Right Side View
https://leetcode.com/problems/binary-tree-right-side-view

Given the root of a binary tree, imagine yourself standing on the right side of it,
return the values of the nodes you can see ordered from top to bottom.

Approaches:
    1. BFS level-order traversal, take the last node of each level.
       Time: O(N), Space: O(N)
"""

from collections import deque
from typing import List, Optional


class TreeNode:
    def __init__(self, v: int):
        self.val = v
        self.left: Optional["TreeNode"] = None
        self.right: Optional["TreeNode"] = None


# Time: O(N), Space: O(N)
def rightSideView(root: Optional[TreeNode]) -> List[int]:
    if not root:
        return []

    res: List[int] = []
    q: deque[TreeNode] = deque()
    q.append(root)

    while q:
        s = len(q)
        for i in range(s):
            curr = q.popleft()

            if i == s - 1:
                res.append(curr.val)

            if curr.left:
                q.append(curr.left)
            if curr.right:
                q.append(curr.right)

    return res


if __name__ == "__main__":
    root1 = TreeNode(1)
    root1.left = TreeNode(2)
    root1.right = TreeNode(3)
    root1.left.right = TreeNode(5)
    root1.right.right = TreeNode(4)
    assert rightSideView(root1) == [1, 3, 4]

    root2 = TreeNode(1)
    root2.right = TreeNode(3)
    assert rightSideView(root2) == [1, 3]

    root3 = None
    assert rightSideView(root3) == []

File: algorithm/binary_tree/cpp/binary_tree_right_side.cpp
// https://leetcode.com/problems/binary-tree-right-side-view

/*
Given the root of a binary tree, imagine yourself standing on the right side of it, return the values of the nodes you can see ordered from top to bottom.

Ex1:
Input: root = [1,2,3,null,5,null,4]
Output: [1,3,4]

Ex2:
Input: root = [1,null,3]
Output: [1,3]

Ex3:
Input: root = []
Output: []

*/

#include <cassert>
#include <vector>
#include <queue>

using namespace std;

class TreeNode {
public:
    int val;
    TreeNode* left;
    TreeNode* right;
    TreeNode(int v) : val(v), left(nullptr), right(nullptr) {}
};

// Time: O(N), Space: O(N)
vector<int> rightSideView(TreeNode* root) {
    if (!root) {
        return {};
    }

    vector<int> res;
    queue<TreeNode*> q;
    q.push(root);

    while (!q.empty()) {
        int s = q.size();
        for (int i = 0; i < s; i++) {
            TreeNode* curr = q.front();
            q.pop();

            if (i == s-1) {
                res.push_back(curr->val);
            }

            if (curr->left) q.push(curr->left);
            if (curr->right) q.push(curr->right);
        }
    }

    return res;
}

int main() {
    TreeNode* root1 = new TreeNode(1);
    root1->left = new TreeNode(2);
    root1->right = new TreeNode(3);
    root1->left->right = new TreeNode(5);
    root1->right->right = new TreeNode(4);
    vector<int> res1 = {1,3,4};
    assert(rightSideView(root1) == res1);

    TreeNode* root2 = new TreeNode(1);
    root2->right = new TreeNode(3);
    vector<int> res2 = {1,3};
    assert(rightSideView(root2) == res2);

    TreeNode* root3 = nullptr;
    vector<int> res3 = {};
    assert(rightSideView(root3) == res3);

    return 0;
}